### DISCUSSION TITLE
How to remove outline-go-tun2socks and Jigsaw-Code/Intra

### DISCUSSION THREAD

**Discussion by AZ-X** [celzero/rethink-app#144] [Ideas]
Google's Jigsaw project is useless. I don't even believe any of authors has watched Saw 3D. 
[outline-go-tun2socks](https://github.com/Jigsaw-Code/outline-go-tun2socks) contains shadowsocks which is fool and I don't want it, instead I want to integrate my own DNS stub written in go [repique](https://github.com/AZ-X/pique). 
BTW, Any of the 2S+ Xposed users may dislike https://github.com/Jigsaw-Code/Intra/

**Comment by ignoramous** [comment]
Checkout AdGuard's DNS Proxy which is a multi protocol DNS client written in Go: https://github.com/AdguardTeam/dnsproxy

As for removing Jigsaw's code is concerned, I don't quite get your concern. Sure we have hit some nasty bugs but it is a pretty good software to build a TCP/UDP firewall on top of, nevertheless.

**Comment by AZ-X** [comment]
> Checkout AdGuard's DNS Proxy which is a multi protocol DNS client written in Go: https://github.com/AdguardTeam/dnsproxy
> 
> As for removing Jigsaw's code is concerned, I don't quite get your concern. Sure we have hit some nasty bugs but it is a pretty good software to build a TCP/UDP firewall on top of, nevertheless.

Adguard is a huge sized program to fulfill its need, but not mine. 

Adguard won't support nested proxy(socks5+udp, http+s), groups, multiple roles on listeners.

[repique](https://github.com/AZ-X/pique) is the only program in this world written in go targeted small size and efficiency.

Your idea is blinking and especially enveloped the fool shadowsocks inside.  However people know things much better than shadowsocks. Why wrapping such useless tool set? 

The rethink for Android is just an out of box DNS stub with GUI written in Java. Am I right? 

> RethinkDNS doesn't aim to be a feature-rich traditional firewall: It is more in-line with Little Snitch than IP tables, say.

**Comment by AZ-X** [comment]
BTW, with [repique](https://github.com/AZ-X/pique), connections to DoH or DoT are limited to TLS 1.3 and certain ciphers.

I don't think it is possible with Adguard or dnsCRYpt-proxy.

**Comment by ignoramous** [comment]
Sorry, I don't get your concern. Do you want us to replace `outline-go-tun2socks` with `repique`? We do need a userspace TCP/IP implementation due to the way the firewall is implemented, and so, another viable option we are looking at is Google's NetStack: #89 `repique` doesn't quite fit the bill.

As for shadowsocks, we do bundle it with the app right now, but it isn't exposed via the app.

**Comment by AZ-X** [comment]
> Sorry, I don't get your concern. Do you want us to replace `outline-go-tun2socks` with `repique`? We do need a userspace TCP/IP implementation due to the way the firewall is implemented, and so, another viable option we are looking at is Google's NetStack: #89 `repique` doesn't quite fit the bill.
> 
> As for shadowsocks, we do bundle it with the app right now, but it isn't exposed via the app.


I thought you have put AdGuard's DNS Proxy in the first place. 

> Checkout AdGuard's DNS Proxy which is a multi protocol DNS client written in Go: https://github.com/AdguardTeam/dnsproxy
> 
> As for removing Jigsaw's code is concerned, I don't quite get your concern. Sure we have hit some nasty bugs but it is a pretty good software to build a TCP/UDP firewall on top of, nevertheless.


I have no idea why want you to remove or replace anything. The issue title is 'How to remove outline-go-tun2socks and Jigsaw-Code/Intra'. I'd like to do it by myself if the whole rethink is capable of plugin and unplug-in.  I want your help to make documentation well done and reconsider adopting anything better than Jigsaw's to fit the need of such as TCP/IP binder. 

> In other words, RethinkDNS has two primary modes, DNS and Firewall. 
> The codebase is raw and is lacking documentation and comprehensive tests. If you need help, feel free to create a Wikipage to highlight the pain with building, testing, writing, committing code.

I care about DNS mode in this program.

**Comment by ignoramous** [comment]
Gotcha. I'm in the process of refactoring the DNS portions of the app and may be once that's done, it'd be easier for you to fork it then. Currently, due to support for multiple DNS modes (Plain DNS, DNS over HTTPS, DNSCrypt) and firewall, the codebase isn't fit for a plug and play.
